src/xc_integrator/replicated/device/incore_replicated_xc_device_integrator_exc_vxc.hpp

```diff
@@ -225,95 +412,84 @@ void IncoreReplicatedXCDeviceIntegrator<ValueType>::
       device_data.generate_buffers( enabled_terms, basis_map, task_it, task_end );
 
     /*** Process the batches ***/
-
+    
     // Evaluate collocation
     if( func.is_gga() ) lwd->eval_collocation_gradient( &device_data );
     else                lwd->eval_collocation( &device_data );
-
-    // Evaluate X matrix
-    lwd->eval_xmat( 2.0, &device_data );
-
-    // Evaluate U/V variables
-    if( func.is_gga() ) lwd->eval_uvvar_gga_rks( &device_data );
-    else                lwd->eval_uvvar_lda_rks( &device_data );
+    
+    auto* data = dynamic_cast<XCDeviceAoSData*>(&device_data);
+    auto tasks = data->host_device_tasks;
+    auto& task = tasks[0];
+    auto static_stack = data->static_stack;
+    auto base_stack   = data->base_stack;
+    
+    const double xmat_fac = is_rks ? 2.0 : 1.0;
+
+    // Evaluate X matrix and V vars
+    lwd->eval_xmat( xmat_fac, &device_data, false, DEN_S );
+    lwd->eval_vvar( &device_data, func.is_gga(), DEN_S );
+
+    if (not is_rks) {
+      lwd->eval_xmat( xmat_fac, &device_data, false, DEN_Z );
+      lwd->eval_vvar( &device_data, func.is_gga(), DEN_Z );
+      if (not is_uks) {
+        lwd->eval_xmat( xmat_fac, &device_data, false, DEN_Y );
+        lwd->eval_vvar( &device_data, func.is_gga(), DEN_Y );
+        lwd->eval_xmat( xmat_fac, &device_data, false, DEN_X );
+        lwd->eval_vvar( &device_data, func.is_gga(), DEN_X );
+      }
+    }
+
+
+    // Evaluate U variables
+    if( func.is_gga() ) lwd->eval_uvars_gga( &device_data, enabled_terms.ks_scheme );
+    else                lwd->eval_uvars_lda( &device_data, enabled_terms.ks_scheme );
 
     // Evaluate XC functional
     if( func.is_gga() ) lwd->eval_kern_exc_vxc_gga( func, &device_data );
     else                lwd->eval_kern_exc_vxc_lda( func, &device_data );
+    
 
     // Do scalar EXC/N_EL integrations
     lwd->inc_exc( &device_data );
     lwd->inc_nel( &device_data );
 
-    // Evaluate Z matrix
-    if( func.is_gga() ) lwd->eval_zmat_gga_vxc_rks( &device_data );
-    else                lwd->eval_zmat_lda_vxc_rks( &device_data );
-
-    // Increment VXC (LT)
-    lwd->inc_vxc( &device_data );
-
-  } // Loop over batches of batches 
-
-  // Symmetrize VXC in device memory
-  lwd->symmetrize_vxc( &device_data );
-
-}
+    
+    if( func.is_gga() ) lwd->eval_zmat_gga_vxc( &device_data, enabled_terms.ks_scheme, DEN_S );
```

Comment:
This could be cleaned up with a lambda
  ```
  auto do_zmat_vxc = [&](den_id) {
    if( func.is_gga() ) 
      lwd->eval_zmat_gga_vxc( &device_data, enabled_terms.ks_scheme, den_id );
    else 
      lwd->eval_zmat_lda_vxc( &device_data, enabled_terms.ks_scheme, den_id );
    lwd->inc_vxc( &device_data, den_id );
  };
  
  do_zmat_vxc(DEN_S);
  if(not is_rks) {
    do_zmat_vxc(DEN_Z);
    if(not is_uks) {
      do_zmat_vxc(DEN_Y);
      do_zmat_vxc(DEN_X);
    }
  }
```